File: Week-Wise-Assignments/Assignment 5/Compressed Solutions/Python Programs/1_double_linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = self.tail = None
    
    def display(self):
        if not self.head: print("Empty"); return
        temp = self.head
        while temp:
            print(temp.data, end=" ")
            temp = temp.next
        print()
    
    def insert_begin(self, data):
        node = Node(data)
        if self.head: self.head.prev = node
        else: self.tail = node
        node.next = self.head
        self.head = node
    
    def insert_end(self, data):
        node = Node(data)
        if self.tail: self.tail.next = node
        else: self.head = node
        node.prev = self.tail
        self.tail = node
    
    def insert_pos(self, data, pos):
        if pos == 0: self.insert_begin(data); return
        node = Node(data)
        temp = self.head
        for _ in range(pos):
            if not temp: return
            temp = temp.next
        if not temp: self.insert_end(data); return
        node.next = temp
        node.prev = temp.prev
        temp.prev.next = node
        temp.prev = node
    
    def search(self, key):
        temp, pos = self.head, 0
        while temp:
            if temp.data == key: return pos
            temp, pos = temp.next, pos + 1
        return -1
    
    def count(self):
        c, temp = 0, self.head
        while temp: c, temp = c + 1, temp.next
        return c
    
    def delete_begin(self):
        if not self.head: return
        if self.head == self.tail: self.head = self.tail = None
        else: self.head, self.head.prev = self.head.next, None
    
    def delete_end(self):
        if not self.tail: return
        if self.head == self.tail: self.head = self.tail = None
        else: self.tail, self.tail.next = self.tail.prev, None
    
    def delete_pos(self, pos):
        if pos == 0: self.delete_begin(); return
        temp = self.head
        for _ in range(pos):
            if not temp: return
            temp = temp.next
        if not temp: return
        if temp.prev: temp.prev.next = temp.next
        if temp.next: temp.next.prev = temp.prev
        if temp == self.tail: self.tail = temp.prev

dll = DoublyLinkedList()
while True:
    choice = int(input("1.Display 2.InsertBegin 3.InsertEnd 4.InsertPos 5.Search 6.Count 7.DelBegin 8.DelEnd 9.DelPos 10.Exit: "))
    if choice == 1: dll.display()
    elif choice == 2: dll.insert_begin(int(input("Data: ")))
    elif choice == 3: dll.insert_end(int(input("Data: ")))
    elif choice == 4: dll.insert_pos(int(input("Data: ")), int(input("Pos: ")))
    elif choice == 5: print(f"Found at: {dll.search(int(input('Key: ')))}")
    elif choice == 6: print(f"Count: {dll.count()}")
    elif choice == 7: dll.delete_begin()
    elif choice == 8: dll.delete_end()
    elif choice == 9: dll.delete_pos(int(input("Pos: ")))
    elif choice == 10: break

File: Week-Wise-Assignments/Assignment 5/Compressed Solutions/C Programs/1_double_linked_list.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
    struct Node* prev;
};

struct Node* head = NULL;

struct Node* createNode(int data) {
    struct Node* node = malloc(sizeof(struct Node));
    node->data = data;
    node->next = node->prev = NULL;
    return node;
}

void display() {
    if (!head) { printf("Empty\n"); return; }
    for (struct Node* temp = head; temp; temp = temp->next)
        printf("%d ", temp->data);
    printf("\n");
}

void insertBegin(int data) {
    struct Node* node = createNode(data);
    if (head) head->prev = node;
    node->next = head;
    head = node;
}

void insertEnd(int data) {
    struct Node* node = createNode(data);
    if (!head) { head = node; return; }
    struct Node* temp = head;
    while (temp->next) temp = temp->next;
    temp->next = node;
    node->prev = temp;
}

void insertPos(int data, int pos) {
    if (pos == 0) { insertBegin(data); return; }
    struct Node* node = createNode(data);
    struct Node* temp = head;
    for (int i = 0; i < pos && temp; i++) temp = temp->next;
    if (!temp) return;
    node->next = temp;
    node->prev = temp->prev;
    if (temp->prev) temp->prev->next = node;
    temp->prev = node;
}

int search(int key) {
    int pos = 0;
    for (struct Node* temp = head; temp; temp = temp->next, pos++)
        if (temp->data == key) return pos;
    return -1;
}

int count() {
    int c = 0;
    for (struct Node* temp = head; temp; temp = temp->next) c++;
    return c;
}

void deleteBegin() {
    if (!head) return;
    struct Node* temp = head;
    head = head->next;
    if (head) head->prev = NULL;
    free(temp);
}

void deleteEnd() {
    if (!head) return;
    struct Node* temp = head;
    while (temp->next) temp = temp->next;
    if (temp->prev) temp->prev->next = NULL;
    else head = NULL;
    free(temp);
}

void deletePos(int pos) {
    if (pos == 0) { deleteBegin(); return; }
    struct Node* temp = head;
    for (int i = 0; i < pos && temp; i++) temp = temp->next;
    if (!temp) return;
    if (temp->prev) temp->prev->next = temp->next;
    if (temp->next) temp->next->prev = temp->prev;
    free(temp);
}

int main() {
    int choice, data, pos;
    while (1) {
        printf("1.Display 2.InsertBegin 3.InsertEnd 4.InsertPos 5.Search 6.Count 7.DelBegin 8.DelEnd 9.DelPos 10.Exit: ");
        scanf("%d", &choice);
        switch (choice) {
            case 1: display(); break;
            case 2: scanf("%d", &data); insertBegin(data); break;
            case 3: scanf("%d", &data); insertEnd(data); break;
            case 4: scanf("%d%d", &data, &pos); insertPos(data, pos); break;
            case 5: scanf("%d", &data); printf("Found at: %d\n", search(data)); break;
            case 6: printf("Count: %d\n", count()); break;
            case 7: deleteBegin(); break;
            case 8: deleteEnd(); break;
            case 9: scanf("%d", &pos); deletePos(pos); break;
            case 10: exit(0);
        }
    }
}